OpenRA.Game/Server/ProtocolVersion.cs

```diff
@@ -36,6 +36,11 @@ public static class ProtocolVersion
 		//   - OrderFields enum encoded as a byte: specifies the data included in the rest of the order
 		//   - Order-specific data - see OpenRA.Game/Server/Order.cs for details
 		//
+		// When the frame of a packet is 0, it is an immediate order, and may or may not be relayed to clients e.g. chat
+		//
+		// When the frame is not 0, it will always be relayed to all clients
+		// and a blank packet sent to the sender to act as an ack (used by adaptive order latency)
```

Comment:
This ack system makes everything considerably more complex and doesn't give any latency advantage. The slightly higher bandwidth usage also shouldn't really make a difference. So why don't you simply send everything through the server without that local loopback special case?

Reply:
That's how I started. I added the ack system on top once it worked without it. It doesn't actually add that much complexity on top of the original code, I'd say most of the complexity is in the server corralling of orders.

Hopefully, we replace the network protocol with UDP as the next step. Despite the `NoDelay` option, TCP still has other issues that can cause unnecessary latency under small amounts of packet loss. There are far more effective ack and bandwidth control schemes for real-time communication that can be written with UDP and will work under very poor conditions. The logic for this stuff will be a lot more complex than the simple loopback acks here, but it will likely supersede it and become an all-in-one protocol that just straight up replaces the current protocol and deciphering logic.

However, this has got me thinking some, so I've got my favourite calculations napkin. This is going to be for the current OpenRA, not this branch: normal game ticks at 40ms and net ticks at 120ms (every 3 frames). Let's assume we have an 8 player game, just for fun.

OpenRA packets have a 12-byte header (length, clientID, frame), which is *per sender*. Each OpenRA packet is also sent on the TCP socket with a single send call, and the TCP socket has NoDelay enabled, which means it also incurs the 40-byte TCP header cost (best case is 20 for IP and 20 for TCP, could be higher with options). Each player sends orders every frame, whether or not they actually did anything because lockstep requires every player receives exact orders from every other player before progressing. This means that each player receives 7 order packets (they don't see their own) every net tick. 7 packets * 8 net ticks * 52 bytes = 2912 bytes/s in pure order overhead (assuming no packet loss).

Every player also sends a sync hash every net frame, and sync hash orders are [13 bytes long](https://github.com/OpenRA/OpenRA/blob/8c9b9df125a047b8d0bdd4c34ab0f221dc82d73f/OpenRA.Game/Network/Order.cs#L53) (includes order type byte), so these packets are 13 + 52 = 65 bytes long. 7 packets * 8 net ticks * 65 bytes = 3640 bytes/s. (Please note that this is also completely redundant since the server checks the sync hashes itself, it only forwards them for simplicity.) So far we've got 6552 bytes/s, which is ~52kbps (more because we rounded down on the number of net ticks per frame). Not great, not terrible, although it will scale proportionally to game speed, so double for the *real* mode, fastest, that apparently nobody plays.

Except that this is all the data required to say "nothing happened, everything is good". We could probably encode all of that data per net tick into less than the 12-byte difference between a single UDP packet and a TCP packet, including custom ack protocol data (4-byte bitfield for redundant ack status of last 32 frames), 1-byte bitfield for clients with orders, 4-byte frame number). We'd have so much to spare that we could up net ticks to every frame, which would actually reduce latency noticeably (to consistently a single game frame if your ping is sub 40ms, which is common nowadays).

Now, orders themselves are currently not very compressed. The order types are string encoded, but probably the worst possible offender is group orders. Each unit in the group is included in the order, and each unit has a 4-byte ID. Let's say we have 8 players with armies of 100 units each move clicking once per second. That's at least 4 bytes per unit * 100 units * 7 players = 2800 bytes/s of just unit IDs! Wait, that's actually not that much... If they were all spam clicking 5 times per second at the same time, it would start to get excessive, sure, but I don't think most players in this game have that high of an APM. It's also possible that they build way more than 100 units each, but seems unlikely to me. There's more data in these orders, such as order strings, target strings, locations, a 1-byte order type, but group move orders are the only type I can think of that would be common enough and large enough to be truly excessive, and the rest of the order will be fairly small in size.

I think this shows just how utterly insane unchecked network overhead could be, we should do some measurements and see if the napkin math checks out.

(Also, the funny thing is that the additional network overhead of my acks is probably greater than the amount saved by emptying those acks instead of just letting the server loopback, although it is still a save since both have that same overhead and we need one of them.)